controllers/masteruserrecord/masteruserrecord_controller.go

```diff
@@ -81,39 +87,98 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
 	if !coputil.IsBeingDeleted(mur) {
 		// Add the finalizer if it is not present
 		if err := r.addFinalizer(logger, mur, murFinalizerName); err != nil {
-			logger.Error(err, "unable to add finalizer to MasterUserRecord")
 			return reconcile.Result{}, err
 		}
 		logger.Info("ensuring user accounts")
-		for _, account := range mur.Spec.UserAccounts {
-			err := r.ensureUserAccount(logger, account, mur)
-			if err != nil {
-				logger.Error(err, "unable to synchronize with member UserAccount")
-				return reconcile.Result{}, err
-			}
+		var membersWithUserAccounts []string
+		if membersWithUserAccounts, err = r.ensureUserAccounts(logger, mur); err != nil || membersWithUserAccounts == nil {
+			return reconcile.Result{}, err
+		}
+		requeueTime, err := r.ensureUserAccountsAreNotPresent(logger, mur, r.membersWithoutUserAccount(membersWithUserAccounts))
+		if err != nil {
+			return reconcile.Result{}, err
+		} else if requeueTime > 0 {
+			return reconcile.Result{RequeueAfter: requeueTime}, err
 		}
-		// If the UserAccount is being deleted, delete the UserAccounts in members.
+
+		// If the MUR is being deleted, delete the UserAccounts in members.
 	} else if coputil.HasFinalizer(mur, murFinalizerName) {
 		requeueTime, err := r.manageCleanUp(logger, mur)
 		if err != nil {
-			logger.Error(err, "unable to clean up MasterUserRecord as part of deletion")
 			return reconcile.Result{}, err
 		} else if requeueTime > 0 {
-			return reconcile.Result{Requeue: true, RequeueAfter: requeueTime}, err
+			return reconcile.Result{RequeueAfter: requeueTime}, err
 		}
 	}
 
 	return reconcile.Result{}, nil
 }
 
+func (r *Reconciler) ensureUserAccountsAreNotPresent(logger logr.Logger, mur *toolchainv1alpha1.MasterUserRecord, targetClusters []string) (time.Duration, error) {
+	for _, toDelete := range targetClusters {
+		requeueTime, err := r.deleteUserAccount(logger, toDelete, mur)
+		if err != nil {
+			return 0, r.wrapErrorWithStatusUpdate(logger, mur, r.setStatusFailed(toolchainv1alpha1.MasterUserRecordUnableToDeleteUserAccountsReason), err,
+				"failed to delete UserAccount in the member cluster '%s'", toDelete)
+		} else if requeueTime > 0 {
+			return requeueTime, nil
+		}
+	}
+	return 0, nil
+}
+
+func (r *Reconciler) membersWithoutUserAccount(membersWithUserAccounts []string) []string {
+	var membersWithout []string
+memberClusters:
+	for memberName := range r.MemberClusters {
+		for _, targetCluster := range membersWithUserAccounts {
+			if memberName == targetCluster {
+				continue memberClusters
+			}
+		}
+		membersWithout = append(membersWithout, memberName)
+	}
+	return membersWithout
+}
+
+func (r *Reconciler) ensureUserAccounts(logger logr.Logger, mur *toolchainv1alpha1.MasterUserRecord) ([]string, error) {
+	spaceBindings := &toolchainv1alpha1.SpaceBindingList{}
+	if err := r.Client.List(context.TODO(), spaceBindings, runtimeclient.InNamespace(mur.GetNamespace())); err != nil {
```

Comment:
Listing all space bindings and then all spaces looks quite concerning from the performance perspective.
Shouldn't we either:
- Just list all the Spaces with the creator label matching the mur owner label and skipping the space binding listing completely
- Or list all the SpaceBindings but with the MUR label matching the MUR name (so we list SBs for the corresponding MUR only vs. all SBs in the system). And then iterate through those SBs to find the "home" Space.

Reply:
I'm idiot, I forgot to add the MUR label :man_facepalming:

Reply:
fixed in [60002c1](https://github.com/codeready-toolchain/host-operator/pull/872/commits/60002c1b61b40e2bb0e23b24707e2ee530bb8864)
Thanks for pointing that out :+1: